### PR TITLE
[improve][fn] support reading config options from file in Function Python Runner

### LINKED ISSUE #18533: [Enhancement] Support reading config options from file in Function Python Runner
### Search before asking

- [X] I searched in the [issues](https://github.com/apache/pulsar/issues) and found nothing similar.

### Motivation

Currently the python runner uses `argparse`  to parse command line args. However in certain cases reading from command line is not very flexible (think we need to use a env var as an argument but the container has no shell access). Thus we need to support reading configs from a file. 

This is necessary for :
- https://github.com/streamnative/function-mesh/issues/368 
- https://github.com/streamnative/function-mesh/issues/448

### Solution

_No response_

### Alternatives

_No response_

### Anything else?

_No response_

### Are you willing to submit a PR?

- [ ] I'm willing to submit a PR!

### PR DESCRIPTION
**Pull request by tpiperatgod**
Fixes #18533

Master Issue: #xyz

### Motivation

please refer to this proposal: https://github.com/apache/pulsar/issues/18744

### Verifying this change

- [ ] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [x] The public API
- [ ] The schema
- [x] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [x] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [x] `doc-required` 
- [ ] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository: https://github.com/tpiperatgod/pulsar/pull/11

### CODE REVIEW

File: pulsar-functions/instance/src/main/python/python_instance_main.py [hunk 1]
```diff
@@ -88,8 +132,10 @@ def main():
   parser.add_argument('--extra_dependency_repository', required=False, help='For packaged python like wheel files, any extra repository to pull the dependencies from')
   parser.add_argument('--state_storage_serviceurl', required=False, help='Managed State Storage Service Url')
   parser.add_argument('--cluster_name', required=True, help='The name of the cluster this instance is running on')
```
Comment (imaffe) [inline pulsar-functions/instance/src/main/python/python_instance_main.py:134]: Should the arugments be declared required ? If a config is provided in file, then maybe the config from command line is not required any more

Comment (tpiperatgod) [reply pulsar-functions/instance/src/main/python/python_instance_main.py]: you're right, I adjust all args to `required=False` and add a new function `validate_arguments()` to verify the merged args, mainly to check the mandatory args.

File: pulsar-functions/instance/src/main/python/python_instance_main.py [hunk 2]
```diff
@@ -49,47 +49,160 @@
...
+  """
+  if config_file is None:
+    return
+  config = util.read_config(config_file)
+  if not config:
+    return
+  default_config = config["DEFAULT"]
+  if not default_config:
+    return
+  if not args.client_auth_plugin and default_config.get("client_auth_plugin", None):
```
Comment (freeznet) [inline pulsar-functions/instance/src/main/python/python_instance_main.py:78]: any chance to simplify this function by [`vars()`](https://docs.python.org/3/library/functions.html#vars)?

**Review by freeznet — COMMENTED** [review]
could you please add some tests?

**Comment by momo-jun** [comment]
Hi @tpiperatgod, did you have any plans to update the related docs?

**Comment by tpiperatgod** [comment]
> Hi @tpiperatgod, did you have any plans to update the related docs?

I will try to submit a PR for the document, please help me review it then, thanks.

**Comment by momo-jun** [comment]
@tpiperatgod I'm happy to help. Feel free to ping me.

**Comment by Anonymitaet** [comment]
Hi @tpiperatgod 3.0 is planned to be released on May 2. We're aiming to align docs with code. Can you share some progress on the doc? TIA!

> > Hi @tpiperatgod, did you have any plans to update the related docs?
> 
> I will try to submit a PR for the document, please help me review it then, thanks.

**Comment by tpiperatgod** [comment]
@Anonymitaet Here is the docs pr, please help review: https://github.com/apache/pulsar-site/pull/541

**Comment by Anonymitaet** [comment]
@tpiperatgod thanks for your technical input! I've restructured the docs based on user learning habits and raised a question in https://github.com/apache/pulsar-site/pull/544#issuecomment-1524712753, PTAL.

**Pull request merged**